alternator/executor.cc

```diff
@@ -3573,6 +3573,10 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
             ++exploded_ck_it;
         }
     }
+    rjson::add_with_string_name(last_evaluated_key, scylla_paging_region, rjson::empty_object());
+    rjson::add(last_evaluated_key[scylla_paging_region.data()], "S", rjson::from_string(to_string(pos.region())));
```

Comment:
> Is this ".data()" neccessary

Yes, `rjson` refuses to work with string views and `operator[]` doesn't have a variant similar to `add_with_string_name` which does work with string view (not that I know of anyway).

Reply:
> The bigger problem I have with this code is that means these new names ":scylla:paging:region" now cannot be used for a key column. This is like #5009 was about the ":attrs" attribute - but actually worse, because 5009 was about fixing the internal representation of a key attribute called ":attrs" - but here you invent a key attribute that can never be used.

I understood based on the Alternator naming rules @psarna linked in earlier in this thread that `:` is not allowed in Alternator column names at all. So it is impossible for a **legal** column name to be named ":scylla:paging:region". Do we enforce this rule?

Reply:
> Is this ".data()" neccessary?
> 
> The bigger problem I have with this code is that means these new names ":scylla:paging:region" now cannot be used for a key column. This is like #5009 was about the ":attrs" attribute - but actually worse, because 5009 was about fixing the internal representation of a key attribute called ":attrs" - but here you invent a key attribute that can never be used.
> 
> Despite this problem I actually think that adding extra attributes to the LastEvaluatedKey is a good idea **if it works** - I also mentioned this idea in #6278 which laments the fact that Alternator has just page positions and not the full cookie like in CQL. I worried that the DynamoDB driver and/or applications might get confused by extra keys in LastEvalutatedKey and either complain or just forget to pass it on to the next request, but it seemed you tried this now and it works. In any case, we'll need to document this difference somewhere - we can be bitten by it down the line.

Yes, I printed the raw JSON before after parsing it and the new fields were present. The driver does seem to do *some* validation, as it enforced the new fields to have the object-value format that the columns have.

Reply:
> > The bigger problem I have with this code is that means these new names ":scylla:paging:region" now cannot be used for a key column. This is like #5009 was about the ":attrs" attribute - but actually worse, because 5009 was about fixing the internal representation of a key attribute called ":attrs" - but here you invent a key attribute that can never be used.
> 
> I understood based on the Alternator naming rules @psarna linked in earlier in this thread that `:` is not allowed in Alternator column names at all. So it is impossible for a **legal** column name to be named ":scylla:paging:region". Do we enforce this rule?

It is allowed according to the link, it's just discouraged, because such names cannot be easily used with bind markers, which start with `:` in DynamoDB. Here's the exact quote (which I already quoted somewhere above, but let me emphasize the important bit:

"""
Although DynamoDB **allows you to use these reserved words and special characters for names**, we recommend that you avoid doing so because you have to define placeholder variables whenever you use these names in an expression. For more information, see [Expression Attribute Names in DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ExpressionAttributeNames.html).
"""

Reply:
So, where should I add a note that field names starting with the `:scylla:` prefix are reserved?

Reply:
I missed the response to this thread, and it looks like this series got merged meanwhile. It means we have a slight regression in alternator, because certain column names have a conflict with the new paging mechanism. I created a post-mortem issue here: https://github.com/scylladb/scylla/issues/10903

Reply:
This irony is that I think eventually we may need to add such extra fake key columns (see https://github.com/scylladb/scylla/issues/6278) but I never understood why we needed them **in this pull request**, and I still don't understand. More about this in 10903, maybe one of you can finally explain it to me :-)